File: numpy/windowing.py
import numpy
from scipy.sparse import random
import sparse
import pytest

from util import RandomScipySparseTensorLoader, RandomPydataSparseTensorLoader

# Want to run these windows so that they operate on
# * A small, constant size grid: 500x500
# * A constant fraction of the tensor size : 1/4 x 1/4
# * Almost the entire tensor (I'm unsure what the point of this comparison is)
# * No windowing (TACO should use a window so we can measure the overhead).
# These options map to the corresponding config values:
sizeConfigs = ["Constant", "ConstantFraction", "AlmostWhole", "Whole", "NoWindowing"]

def sliceTensor(tensor, dim, config):
    if config == "Constant":
        return tensor[250:750, 250:750]
    elif config == "ConstantFraction":
        size = dim//4
        start = dim//4
        return tensor[start:(start+size), start:(start+size)]
    elif config == "AlmostWhole":
        return tensor[1:(dim-1), 1:(dim-1)]
    elif config == "Whole":
        return tensor[0:dim, 0:dim]
    elif config == "NoWindowing":
        return tensor
    else:
        assert(False)

@pytest.mark.parametrize("dim", [5000, 10000, 20000])
@pytest.mark.parametrize("format", ['csr'])
@pytest.mark.parametrize("config", sizeConfigs)
def bench_add_sparse_window(tacoBench, dim, format, config):
    loader = RandomScipySparseTensorLoader(format)
    matrix = loader.random((dim, dim), 0.01).astype('float64')
    matrix2 = loader.random((dim, dim), 0.01, variant=1).astype('float64')
    def bench():
        x = sliceTensor(matrix, dim, config)
        x2 = sliceTensor(matrix2, dim, config)
        res = x + x2
        # Sanity check that this has a similar runtime as taco.
        # res = matrix + matrix
    tacoBench(bench)

# TODO (rohany): This is really slow (compared to scipy.sparse). Check with hameer
#  that this result makes sense.
@pytest.mark.parametrize("dim", [5000, 10000, 20000])
@pytest.mark.parametrize("config", sizeConfigs)
def bench_add_pydata_sparse_window(tacoBench, dim, config):
    loader = RandomPydataSparseTensorLoader()
    matrix = loader.random((dim, dim), 0.01).astype('float64')
    matrix2 = loader.random((dim, dim), 0.01, variant=1).astype('float64')
    def bench():
        x = sliceTensor(matrix, dim, config)
        x2 = sliceTensor(matrix2, dim, config)
        res = x + x2
    tacoBench(bench)

@pytest.mark.parametrize("dim", [5000, 10000, 20000])
@pytest.mark.parametrize("format", ['csr'])
@pytest.mark.parametrize("strideWidth", [2, 4, 8])
def bench_add_sparse_strided_window(tacoBench, dim, format, strideWidth):
    loader = RandomScipySparseTensorLoader(format)
    matrix = loader.random((dim, dim), 0.01).astype('float64')
    matrix2 = loader.random((dim, dim), 0.01, variant=1).astype('float64')
    def bench():
        x = matrix[0:dim:strideWidth, 0:dim:strideWidth] 
        x2 = matrix2[0:dim:strideWidth, 0:dim:strideWidth] 
        res = x + x2
    tacoBench(bench)

@pytest.mark.parametrize("dim", [5000, 10000, 20000])
@pytest.mark.parametrize("format", ['csr', 'csc'])
@pytest.mark.parametrize("fraction", [2, 4, 8])
@pytest.mark.skip(reason="not doing index sets")
def bench_add_sparse_index_set(tacoBench, dim, format, fraction):
    indexes = [i * fraction for i in range(0, dim//fraction)]
    loader = RandomScipySparseTensorLoader(format)
    matrix = loader.random((dim, dim), 0.01)
    matrix2 = loader.random((dim, dim), 0.01, variant=1)
    def bench():
        x = matrix[:, indexes] 
        x2 = matrix2[:, indexes] 
        res = x + x2
    tacoBench(bench)

@pytest.mark.parametrize("dim", [5000, 10000, 20000])
@pytest.mark.parametrize("strideWidth", [2, 4, 8])
def bench_add_pydata_sparse_strided_window(tacoBench, dim, strideWidth):
    loader = RandomPydataSparseTensorLoader()
    matrix = loader.random((dim, dim), 0.01).astype('float64')
    matrix2 = loader.random((dim, dim), 0.01, variant=1).astype('float64')
    def bench():
        x = matrix[0:dim:strideWidth, 0:dim:strideWidth] 
        x2 = matrix2[0:dim:strideWidth, 0:dim:strideWidth] 
        res = x + x2
    tacoBench(bench)

# TODO (rohany): This is really slow (compared to scipy.sparse). Check with hameer
#  that this result makes sense.
@pytest.mark.parametrize("dim", [5000, 10000, 20000])
@pytest.mark.parametrize("fraction", [2, 4, 8])
@pytest.mark.skip(reason="not doing index sets")
def bench_add_pydata_sparse_index_set(tacoBench, dim, fraction):
    loader = RandomPydataSparseTensorLoader()
    indexes = [i * fraction for i in range(0, dim//fraction)]
    matrix = loader.random((dim, dim), 0.01)
    matrix2 = loader.random((dim, dim), 0.01, variant=1)
    def bench():
        x = matrix[:, indexes] 
        x2 = matrix2[:, indexes] 
        res = x + x2
    tacoBench(bench)

@pytest.mark.parametrize("dim", [5000, 10000, 20000])
@pytest.mark.parametrize("format", ['csr', 'csc'])
@pytest.mark.skip(reason="not using currently")
def bench_add_multiple_sparse_windows(tacoBench, dim, format):
    matrix1 = random(dim, dim, format=format)
    matrix2 = random(dim, dim, format=format)
    def bench():
        res = matrix1[1:(dim-1), 1:(dim-1)] + matrix2[1:(dim-1), 1:(dim-1)] + matrix1[0:(dim-2), 0:(dim-2)]
    tacoBench(bench)

@pytest.mark.skip(reason="too slow right now")
def bench_add_window(tacoBench):
    dim = 10000
    matrix = random(dim, dim, format="csr").todense()
    def bench():
        x = matrix[1:(dim-1), 1:(dim-1)]
        res = x + x
    tacoBench(bench)

File: taco/windowing.cpp
#include "bench.h"
#include "benchmark/benchmark.h"

#include "taco/tensor.h"
#include "taco/format.h"
#include "taco/index_notation/index_notation.h"

using namespace taco;

// The tensor sizes that we want to run each windowing benchmark with.
std::vector<int64_t> tensorSizes({5000, 10000, 20000});

// WindowConfig corresponds to what sort of window size should be used
// when evaluating standard windowing benchmarks. This should stay in
// line with the definition in numpy/windowing.py.
enum WindowConfig {
  Constant,
  ConstantFraction,
  AlmostWhole,
  Whole,
  NoWindowing,
};

#define FOREACH_WINDOW_CONFIG(__func__) \
  __func__(Constant, Constant) \
  __func__(ConstantFraction, ConstantFraction) \
  __func__(AlmostWhole, AlmostWhole) \
  __func__(Whole, Whole) \
  __func__(NoWindowing, NoWindowing)

Tensor<double> windowedTensorOp(Tensor<double> input1, Tensor<double> input2, int dim, WindowConfig config) {
  IndexVar i, j;
  switch (config) {
    case Constant: {
      Tensor<double> result("B", {500, 500}, input1.getFormat());
      result(i, j) = input1(i(250, 750), j(250, 750)) + input2(i(250, 750), j(250, 750));
      return result;
    }
    case ConstantFraction: {
      int size = dim / 4;
      int start = dim / 4;
      Tensor<double> result("B", {size, size}, input1.getFormat());
      result(i, j) = input1(i(start, start + size), j(start, start + size)) + input2(i(start, start + size), j(start, start + size));
      return result;
    }
    case AlmostWhole: {
      Tensor<double> result("B", {dim-2, dim-2}, input1.getFormat());
      result(i, j) = input1(i(1, dim-1), j(1, dim-1)) + input2(i(1, dim-1), j(1, dim-1));
      return result;
    }
    case Whole: {
      Tensor<double> result("B", {dim, dim}, input1.getFormat());
      result(i, j) = input1(i(0, dim), j(0, dim)) + input2(i(0, dim), j(0, dim));
      return result;
    }
    case NoWindowing: {
      Tensor<double> result("B", {dim, dim}, input1.getFormat());
      result(i, j) = input1(i, j) + input2(i, j);
      return result;
    }
    default:
      assert(false);
  }
}

static void bench_add_sparse_window(benchmark::State& state, const Format& f, WindowConfig config) {
  int dim = state.range(0);
  auto sparsity = 0.01;
  Tensor<double> matrix = loadRandomTensor("A", {dim, dim}, sparsity, f);
  Tensor<double> matrix2 = loadRandomTensor("A2", {dim, dim}, sparsity, f, 1 /* variant */);
  matrix.pack();

  for (auto _ : state) {
    // Setup.
    state.PauseTiming();
    auto result = windowedTensorOp(matrix, matrix2, dim, config);
    result.setAssembleWhileCompute(true);
    result.compile();
    state.ResumeTiming();
    // The actual computation.
    result.compute();
  }
}

#define DECLARE_ADD_SPARSE_WINDOW_BENCH(configName, config) \
  TACO_BENCH_ARGS(bench_add_sparse_window, csr/configName, CSR, config)->ArgsProduct({tensorSizes});

FOREACH_WINDOW_CONFIG(DECLARE_ADD_SPARSE_WINDOW_BENCH)

static void bench_add_sparse_strided_window(benchmark::State& state, const Format& f) {
  int dim = state.range(0);
  int strideWidth = state.range(1);
  auto sparsity = 0.01;
  Tensor<double> matrix = loadRandomTensor("A", {dim, dim}, sparsity, f);
  Tensor<double> matrix2 = loadRandomTensor("A2", {dim, dim}, sparsity, f, 1 /* variant */);
  matrix.pack();

  for (auto _ : state) {
    // Setup.
    state.PauseTiming();
    Tensor<double> result("B", {dim/strideWidth, dim/strideWidth}, f);
    IndexVar i, j;
    result(i, j) = matrix(i(0, dim, strideWidth), j(0, dim, strideWidth)) + matrix2(i(0, dim, strideWidth), j(0, dim, strideWidth));
    result.setAssembleWhileCompute(true);
    result.compile();
    state.ResumeTiming();
    // The actual computation.
    result.compute();
  }
}
std::vector<int64_t> strides({2, 4, 8});
TACO_BENCH_ARG(bench_add_sparse_strided_window, csr, CSR)
  ->ArgsProduct({tensorSizes, strides});
TACO_BENCH_ARG(bench_add_sparse_strided_window, csc, CSC)
  ->ArgsProduct({tensorSizes, strides});

static void bench_add_sparse_index_set(benchmark::State& state, const Format& f) {
  int dim = state.range(0);
  int fraction = state.range(1);
  auto sparsity = 0.01;
  Tensor<double> matrix = loadRandomTensor("A", {dim, dim}, sparsity, f);
  Tensor<double> matrix2 = loadRandomTensor("A2", {dim, dim}, sparsity, f, 1 /* variant */);
  std::vector<int> indexSet;
  for (int i = 0; i < dim / fraction; i++) {
    indexSet.push_back(i * fraction);
  }

  for (auto _ : state) {
    // Setup.
    state.PauseTiming();
    Tensor<double> result("B", {dim, dim / fraction}, f);
    IndexVar i("i"), j("j");
    result(i, j) = matrix(i, j(indexSet)) + matrix2(i, j(indexSet));
    result.setAssembleWhileCompute(true);
    result.compile();
    state.ResumeTiming();

    result.compute();
  }
}
// Not doing index sets currently.
// std::vector<int64_t> fractions({2, 4, 8});
// TACO_BENCH_ARG(bench_add_sparse_index_set, csr, CSR)
//   ->ArgsProduct({tensorSizes, fractions});
